### ISSUE TITLE
Perform static argument transformation

### ISSUE THREAD

**Issue by phischu** [effekt-lang/effekt#690]
Consider the following program:

```
interface Fail {
  def fail(): Nothing
}

def choice(n: Int): Int / Fail =
  if (n < 1) {
    do fail() match {}
  } else {
    choice(n - 1)
  }

def run(n: Int) =
  try {
    val i = choice(n)
    val j = choice(i - 1)
    (i + j)
  } with Fail {
    def fail() = 0
  }

def main() = println(run(10))
```

I contains a loop `choice` which can not and will not be inlined. After optimization we produce the following Core:

```
interface Fail {
  fail: () => Nothing
}

def choice(n: Int){Fail} =
  if (infixLt(n, 1)) {
    val v_r: Nothing = Fail.fail();
    v_r match {
    }
  } else {
    choice(infixSub(n, 1), Fail)
  }

def main() =
  let v_r = run {reset { (){p} =>
    val i: Int = choice(10, new Fail {
      def fail() =
        shift(p) { (){k} => 0 }
    });
    val j: Int = choice(infixSub(i, 1), new Fail {
      def fail() =
        shift(p) { (){k} => 0 }
    });
    infixAdd(i, j)
  }};
  let v_r = println1(show(v_r))
  v_r
```

I would like to see some static argument transformation, which in general is probably very difficult. The recursive function `choice` should be specialized to its two call-sites:

```
interface Fail {
  fail: () => Nothing
}

def choice_1(n: Int, p: Prompt) =
  if (infixLt(n, 1)) {
    val v_r: Nothing = shift(p) { (){k} => 0 };
    v_r match {
    }
  } else {
    choice_1(infixSub(n, 1), p)
  }

def choice_2(n: Int, p: Prompt) =
  if (infixLt(n, 1)) {
    val v_r: Nothing = shift(p) { (){k} => 0 };
    v_r match {
    }
  } else {
    choice_2(infixSub(n, 1), p)
  }

def main() =
  let v_r = run {reset { (){p} =>
    val i: Int = choice_1(10, p);
    val j: Int = choice_2(infixSub(i, 1), p);
    infixAdd(i, j)
  }};
  let v_r = println1(show(v_r))
  v_r
```

My expectation is that performing a static argument transformation will yield great speed-ups in effect-handler-heavy benchmarks.

**Comment by marvinborner** [comment]
Is this issue resolved by #700? To be honest I don't really see the benefit of specializing your example into the two callsites. Right now, your example gets optimized to the following core (before inlining):

``` scala
def choice(_n: Int){Fail} =
  def choice_worker(n: Int) =
    if (infixLt(n, 1)) {
      val v_r: Nothing = Fail.fail();
      v_r match {
      }
    } else {
      choice_worker(infixSub(n, 1))
    }
  choice_worker(_n);

def run(n: Int) =
  reset { (){p} =>
    def Fail = new Fail {
      def fail() = shift(p) { (){k} => 0 }
    }

    val i: Int = choice(n, Fail);
    val j: Int = choice(infixSub(i, 1), Fail);
    infixAdd93(i, j)
  };

def main() =
  let v_r = run {run(10)}
  println(v_r)
```

**Comment by b-studios** [comment]
I think this is resolved for now. The two callsites are easier since we do not need to remember that we have seen a similar specialization. Inlining will currently do exactly that. In the future, we can now try to be smarter and not duplicate it.
